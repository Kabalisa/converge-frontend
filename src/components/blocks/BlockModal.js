import PropTypes from 'prop-types';
import React from 'react';
import Modal from '../commons/Modal';
import { Input } from '../commons/index';
import ActionButtons from '../commons/ActionButtons';

const BlockModal = props => (
  <Modal
    title={!props.editing ? props.office.name : 'EDIT Block'}
    buttonText={!props.editing ? props.office.name : 'Edit'}
    closeModal={props.closeModal}
    handleCloseRequest={props.handleModalStateChange}
    className="add-office-modal"
    // modalButton="add-button"
  >
    <div className="modal-form">
      <Input
        labelName="Block Name"
        name="blockName"
        value={props.blockName}
        placeholder="Enter block name"
        id="blockName"
        onChange={props.handleFormInputChange}
      />
      <ActionButtons
        withCancel
        isLoading={props.isLoading}
        onClickCancel={props.handleCloseModal}
        onClickSubmit={props.handleSubmit}
        actionButtonText={props.editing ? 'SAVE CHANGES' : 'ADD BLOCK'}
      />
    </div>
  </Modal>
);

BlockModal.propTypes = {
  handleModalStateChange: PropTypes.func.isRequired,
  closeModal: PropTypes.bool.isRequired,
  handleCloseModal: PropTypes.func.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  handleFormInputChange: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
  office: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    location: PropTypes.shape({
      name: PropTypes.string,
    }),
  }).isRequired,
  blockName: PropTypes.string,
  editing: PropTypes.bool.isRequired,
};

BlockModal.defaultProps = {
  blockName: '',
  isLoading: false,
};

export default BlockModal;
