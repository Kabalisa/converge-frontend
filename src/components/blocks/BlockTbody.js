import React from 'react';
import PropTypes from 'prop-types';
import SingleBlock from './SingleBlockDetails';

const formatBlockData = block => ({
  id: block.id,
  name: block.name,
  office: block.offices,
  location: block.offices.location.name,
});

const BlockTableBody = props => (
  <tbody className="bundle">
    {props.blocks.length > 0 && props.blocks.map(block => (
      <SingleBlock
        block={formatBlockData(block)}
        key={block.id}
        refetch={props.refetch}
      />
  ))}
  </tbody>
);

BlockTableBody.propTypes = {
  blocks: PropTypes.arrayOf(PropTypes.object).isRequired,
  refetch: PropTypes.func,
};

BlockTableBody.defaultProps = {
  refetch: null,
};

export default BlockTableBody;
