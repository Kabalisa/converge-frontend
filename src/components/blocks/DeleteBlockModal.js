import PropTypes from 'prop-types';
import React from 'react';
import Modal from '../commons/Modal';
import ActionButtons from '../commons/ActionButtons';

const DeleteBlockModal = props => (
  <Modal
    handleCloseRequest={props.handleModalStateChange}
    title="DELETE Block"
    buttonText="Delete"
    className="add-office-modal"
    closeModal={props.closeModal}
  >
    <div className="delete-modal-content">
      <p id="confirm-msg">
        Are you sure you want to delete {`"${props.block.name}"`}? <br />
        This cannot be undone
      </p>
      <ActionButtons
        withCancel
        onClickCancel={props.handleCloseModal}
        isLoading={props.isLoading}
        actionButtonText="DELETE BLOCK"
        onClickSubmit={props.handleDeleteBlock}
      />
    </div>
  </Modal>
);

DeleteBlockModal.propTypes = {
  handleModalStateChange: PropTypes.func.isRequired,
  closeModal: PropTypes.bool.isRequired,
  handleCloseModal: PropTypes.func.isRequired,
  handleDeleteBlock: PropTypes.func.isRequired,
  isLoading: PropTypes.bool,
  block: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
  }).isRequired,
};

DeleteBlockModal.defaultProps = {
  isLoading: false,
};

export default DeleteBlockModal;
