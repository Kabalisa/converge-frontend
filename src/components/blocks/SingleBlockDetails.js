/* eslint-disable import/no-named-as-default */
import React from 'react';
import PropTypes from 'prop-types';
import Block from './BlockActions';

const SingleBlockDetails = (props) => {
  const { name, office, location } = props.block;
  return (
    <div className="table__row">
      <span>{name}</span>
      <span>{location}</span>
      <span>{office.name}</span>
      <span>
        <Block office={office} block={props.block} editing /> &nbsp;
        <Block block={props.block} office={office} refetch={props.refetch} deleting /> &nbsp;
      </span>
    </div>
  );
};

SingleBlockDetails.propTypes = {
  block: PropTypes.shape({
    name: PropTypes.string.isRequired,
    office: PropTypes.shape({
      id: PropTypes.string,
      name: PropTypes.string,
      location: PropTypes.shape({
        name: PropTypes.string,
      }),
    }).isRequired,
  }).isRequired,
  refetch: PropTypes.func,
};

SingleBlockDetails.defaultProps = {
  refetch: null,
};

export default SingleBlockDetails;
