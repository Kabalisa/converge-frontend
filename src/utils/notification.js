/* eslint-disable */
/**
 * toastr => This is the toastr module imported from 'toastr' npm package
 * notifyType => This is the notification type (success or error)
 * message => This is the notification message
 *
 * How to use this function
 * const notify = notification(toastr, 'success', 'noficationMessage');
 * notify()
 */
const notification = (toastr, notifyType, message) => {
  toastr['options'] = {
    'closeButton': true,
    'positionClass': 'toast-top-center',
    'preventDuplicates': true,
    'onclick': null,
    'showDuration': '300',
    'hideDuration': '1000',
    'timeOut': '5000',
    'extendedTimeOut': '100000',
    'showEasing': 'swing',
    'hideEasing': 'linear',
    'showMethod': 'fadeIn',
    'hideMethod': 'fadeOut',
  };
  return () => toastr[notifyType](message);
};

export default notification;