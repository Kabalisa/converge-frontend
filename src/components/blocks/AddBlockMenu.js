import React from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import { IconMenu, MenuItem, MenuDivider } from 'react-toolbox/lib/menu';
import '../../assets/styles/addRoomMenu.scss';
import AddBlock from './BlockActions';
import { decodeTokenAndGetUserData } from '../../utils/Cookie';
import { GET_USER_QUERY } from '../../graphql/queries/People';

const menuText = () => (
  <div className="addBlockBtn">
    <span>Add Block</span>
  </div>
);

export const AddBlockMenu = (props) => {
  const { offices: { offices }, user: { loading, user }, refetch } = props;
  return (
    <IconMenu className="add-block-menu" icon={menuText()}>
      <MenuItem caption="Select Office" disabled />
      <MenuDivider />
      <div className="add-block-menu-list">
        {(offices && !loading) &&
          offices.filter(office => office.location.name === user.location)
          .map(office => (
            <MenuItem key={office.id}>
              <AddBlock office={office} refetch={refetch} />
            </MenuItem>
          ),
        )}
      </div>
    </IconMenu>
  );
};

/* This gets the token from the localstorage and select the user
email to pass as a parameter to the query being sent */
const { UserInfo: userData } = decodeTokenAndGetUserData() || {};

AddBlockMenu.propTypes = {
  offices: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    location: PropTypes.shape({
      name: PropTypes.string,
    }),
  }).isRequired,
  refetch: PropTypes.func,
  user: PropTypes.shape({
    location: PropTypes.string,
  }),
};

AddBlockMenu.defaultProps = {
  refetch: null,
  user: { location: 'Nairobi' },
};

export default compose(
  graphql(GET_USER_QUERY, {
    name: 'user',
    options: /* istanbul ignore next */ () => ({
      variables: {
        email: userData.email,
      },
    }),
  }),
)(AddBlockMenu);
