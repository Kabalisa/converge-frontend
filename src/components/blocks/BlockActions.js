import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { graphql, compose } from 'react-apollo';
import toastr from 'toastr';
import '../../assets/styles/addoffice.scss';
import ADD_BLOCK_MUTATION, { DELETE_BLOCK_MUTATION, EDIT_BLOCK_MUTATION } from '../../graphql/mutations/Blocks';
import notification from '../../utils/notification';
import BlockModal from './BlockModal';
import DeleteBlockModal from './DeleteBlockModal';

export class BlockActions extends Component {
  state = {
    blockName: this.props.block.name,
    closeModal: false,
    isLoading: false,
  };

  handleCloseModal = () => {
    this.setState({ closeModal: true });
  };

  handleFormInputChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  };

  handleModalStateChange = () => {
    this.state.closeModal && this.setState({ closeModal: false });
  };

  handleAddBlock = (event) => {
    event.preventDefault();
    const { blockName } = this.state;
    const { office, refetch } = this.props;
    if (!blockName) {
      notification(toastr, 'error', 'SingleBlock name is required')();
    } else {
      this.toggleLoading();
      this.props
        .addBlock({
          variables: {
            officeId: office.id,
            name: blockName,
          },
        })
        .then(() => {
          this.toggleLoading();
          this.handleCloseModal();
          refetch();
          notification(
            toastr,
            'success',
            `${blockName} block has been added successfully`,
          )();
        })
        .catch((err) => {
          this.toggleLoading();
          this.handleCloseModal();
          notification(toastr, 'error', err.graphQLErrors[0].message)();
        });
    }
  };

  handleDeleteBlock = (event) => {
    event.preventDefault();
    const { block, refetch } = this.props;
    if (!block || Object.keys(block).length === 0) {
      notification(toastr, 'error', 'SingleBlock not specified is required')();
    } else {
      this.toggleLoading();
      this.props
        .deleteBlock({
          variables: {
            blockId: block.id,
          },
        })
        .then(() => {
          this.toggleLoading();
          this.handleCloseModal();
          refetch();
          notification(
            toastr,
            'success',
            `block ${block.name} has been deleted successfully`,
          )();
        })
        .catch((err) => {
          this.toggleLoading();
          this.handleCloseModal();
          notification(toastr, 'error', err.graphQLErrors[0].message)();
        });
    }
  };

  handleEditBlock = (event) => {
    event.preventDefault();
    const { block } = this.props;
    const { blockName } = this.state;
    if (!blockName) {
      notification(toastr, 'error', 'SingleBlock name is required')();
    } else {
      this.toggleLoading();
      this.props
        .editBlock({
          variables: {
            blockId: block.id,
            name: blockName,
          },
        })
        .then(() => {
          this.toggleLoading();
          this.handleCloseModal();
          notification(
            toastr,
            'success',
            `block has been renamed successfully to ${blockName}`,
          )();
        })
        .catch((err) => {
          this.toggleLoading();
          this.handleCloseModal();
          notification(toastr, 'error', err.graphQLErrors[0].message)();
        });
    }
  };

  /**
   * 1. change isLoading state to it's opposite value
   * i.e true to false or vise verser
   *
   * @returns {void}
   */
  toggleLoading = () => {
    this.setState({
      isLoading: !this.state.isLoading,
    });
  }

  render() {
    const { blockName, closeModal, isLoading } = this.state;
    const {
      office, editing, deleting, block,
    } = this.props;
    if (deleting) {
      return (
        <DeleteBlockModal
          block={block}
          closeModal={closeModal}
          isLoading={isLoading}
          handleCloseModal={this.handleCloseModal}
          handleModalStateChange={this.handleModalStateChange}
          handleDeleteBlock={this.handleDeleteBlock}
        />);
    }
    return (
      <BlockModal
        blockName={blockName}
        closeModal={closeModal}
        isLoading={isLoading}
        office={editing ? {} : office}
        editing={editing}
        handleCloseModal={this.handleCloseModal}
        handleSubmit={editing ? this.handleEditBlock : this.handleAddBlock}
        handleFormInputChange={this.handleFormInputChange}
        handleModalStateChange={this.handleModalStateChange}
      />);
  }
}

BlockActions.propTypes = {
  office: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    location: PropTypes.shape({
      name: PropTypes.string,
    }),
  }).isRequired,
  editing: PropTypes.bool,
  deleting: PropTypes.bool,
  block: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
  }),
  editBlock: PropTypes.func.isRequired,
  addBlock: PropTypes.func.isRequired,
  deleteBlock: PropTypes.func.isRequired,
  refetch: PropTypes.func,
};

BlockActions.defaultProps = {
  editing: false,
  deleting: false,
  block: {},
  refetch: () => {},
};

export default compose(
  graphql(ADD_BLOCK_MUTATION, { name: 'addBlock' }),
  graphql(DELETE_BLOCK_MUTATION, { name: 'deleteBlock' }),
  graphql(EDIT_BLOCK_MUTATION, { name: 'editBlock' }),
)(BlockActions);
