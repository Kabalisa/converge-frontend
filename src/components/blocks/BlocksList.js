import React from 'react';
import PropTypes from 'prop-types';
import { compose, graphql } from 'react-apollo';
import Spinner from '../commons/Spinner';
import ColGroup from '../helpers/ColGroup';
import TableHead from '../helpers/TableHead';
import BlockTableBody from './BlockTbody';
import GET_ALL_BLOCKS from '../../graphql/queries/Blocks';
// eslint-disable-next-line import/no-named-as-default
import AddBlockMenu from './AddBlockMenu';
import { GET_ALL_OFFICES } from '../../graphql/queries/Offices';
import MenuTitle from '../commons/MenuTitle';

export const BlocksList = (props) => {
  const { allOffices } = props.allOffices;
  const { allBlocks, loading, refetch } = props.allBlocks;
  if (allBlocks === undefined || allOffices === undefined || loading) return <Spinner />;

  return (
    <div className="settings-rooms">
      <div className="settings-rooms-control">
        <MenuTitle title="Blocks" />
        <AddBlockMenu offices={allOffices} refetch={refetch} />
      </div>
      <div className="settings-rooms-list">
        <table>
          <ColGroup />
          <TableHead titles={['SingleBlock', 'Location', 'Office', 'Action']} />
          <BlockTableBody blocks={allBlocks} refetch={refetch} />
        </table>
      </div>
    </div>
  );
};

BlocksList.propTypes = {
  allOffices: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    location: PropTypes.shape({
      name: PropTypes.string,
    }),
  }).isRequired,
  allBlocks: PropTypes.shape({
    id: PropTypes.number,
    name: PropTypes.string,
    offices: PropTypes.shape({
      name: PropTypes.string,
    }),
  }).isRequired,
};

export default compose(
  graphql(GET_ALL_BLOCKS, {
    name: 'allBlocks',
  }),
  graphql(GET_ALL_OFFICES,
    {
      name: 'allOffices',
      options: () => ({
        variables: {
          page: 1,
          perPage: 1000,
        },
      }),
    }),
)(BlocksList);
