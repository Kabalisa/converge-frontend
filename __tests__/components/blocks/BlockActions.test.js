import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloError } from 'apollo-client';
import { BlockActions } from '../../../src/components/blocks/BlockActions';
import blockData from '../../../src/fixtures/blocksData';
import officeData from '../../../src/fixtures/officeData';
import notification from '../../../src/utils/notification';

const props = {
  office: officeData,
  block: blockData,
  addBlock: jest.fn(),
  editBlock: jest.fn(),
  deleteBlock: jest.fn(),
};
const event = {
  preventDefault: jest.fn(),
  target: {
    name: '',
    value: '',
  },
};

describe('BlockActions Component', () => {
  let wrapper;
  let instance;
  let mountedComponent;
  let mountedInstance;

  beforeEach(() => {
    wrapper = shallow(<BlockActions {...props} />);
    instance = wrapper.instance();
    jest.spyOn(instance, 'handleCloseModal');
    jest.spyOn(instance, 'handleDeleteBlock');
    jest.spyOn(instance, 'handleFormInputChange');
    jest.spyOn(instance, 'handleModalStateChange');
    jest.spyOn(instance, 'toggleLoading');

    mountedComponent = mount(<BlockActions {...props} />);
    mountedInstance = mountedComponent.instance();
    jest.spyOn(mountedInstance, 'handleModalStateChange');
  });

  it('renders properly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders div modal', () => {
    const modal = wrapper.dive().find('.modal-form');
    expect(modal).toHaveLength(1);
  });

  it('should set blockName to state on click', () => {
    const toTarget = { target: { name: 'blockName', value: 'block Test' } };
    wrapper
      .dive()
      .find('#blockName')
      .simulate('change', toTarget);
    expect(wrapper.state('blockName')).toBe('block Test');
  });

  it('should update closeModal state', () => {
    wrapper.instance().handleCloseModal();
    expect(wrapper.state('closeModal')).toEqual(true);
  });

  it('should close modal', () => {
    wrapper.setState({
      closeModal: true,
    });
    wrapper.update();
    wrapper.instance().handleModalStateChange();
    expect(wrapper.state('closeModal')).toEqual(false);
  });

  it('should render Editing Modal', () => {
    wrapper.setProps({ editing: true });
    wrapper.update();
    const form = wrapper.dive().find('.modal-form');
    expect(form).toHaveLength(1);
  });

  it('should render Deleting modal', () => {
    wrapper.setProps({ deleting: true });
    wrapper.update();
    const blockName = wrapper.dive().find('[title="DELETE Block"]');
    expect(blockName).toHaveLength(1);
  });

  it('should close the modal when the addBlock call fails and blockName is not provided', () => {
    wrapper.setProps({
      addBlock: jest.fn(() => Promise.reject(new ApolloError({ graphQLErrors: [new Error('error')] }))),
      office: {},
    });
    wrapper.setState({
      blockName: '',
    });
    wrapper.update();
    wrapper.instance().handleAddBlock(event);
    expect(wrapper.state('closeModal')).toEqual(false);
  });

  it('should close the modal when the addBlock call fails and blockName is provided', () => {
    wrapper.setProps({
      addBlock: jest.fn(() => Promise.reject(new ApolloError({ graphQLErrors: [new Error('error')] }))),
    });
    wrapper.setState({
      blockName: 'Test SingleBlock',
    });
    wrapper.update();
    wrapper.instance().handleAddBlock(event);
    expect(instance.toggleLoading).toBeCalled();
    expect(wrapper.state('closeModal')).toEqual(false);
  });

  it('should resolve addBlock', () => {
    const newProps = {
      addBlock: jest.fn(() => Promise.resolve()),
      office: officeData,
      block: blockData,
      editBlock: jest.fn(),
      deleteBlock: jest.fn(),
    };
    wrapper = shallow(<BlockActions {...newProps} />);
    wrapper.setState({
      blockName: 'Sample SingleBlock',
    });
    wrapper.update();
    wrapper.instance().handleAddBlock(event);
    expect(newProps.addBlock).toBeCalled();
  });

  it('should close the modal when the editBlock call fails and blockName is not provided', () => {
    wrapper.setProps({
      editBlock: jest.fn(() => Promise.reject().catch(() => {})),
      office: {},
      block: {},
    });
    wrapper.setState({
      blockName: '',
    });
    wrapper.update();
    wrapper.instance().handleEditBlock(event);
    expect(wrapper.state('closeModal')).toEqual(false);
  });

  it('should close the modal when the editBlock call fails and office is not provided', () => {
    notification.default = jest.fn();
    wrapper.setProps({
      editBlock: jest.fn(() => Promise.reject(new ApolloError({ graphQLErrors: [new Error('error')] }))),
      office: {},
    });
    wrapper.setState({
      blockName: 'Test SingleBlock',
    });
    wrapper.update();
    wrapper.instance().handleEditBlock(event);
    expect(instance.toggleLoading).toBeCalled();
    expect(wrapper.state('closeModal')).toEqual(false);
  });

  it('should call editBlock when handleEditBlock is called', () => {
    const newProps = {
      editBlock: jest.fn(() => Promise.resolve()),
      office: officeData,
      block: blockData,
      addBlock: jest.fn(),
      deleteBlock: jest.fn(),
    };
    wrapper = shallow(<BlockActions {...newProps} />);
    wrapper.setState({
      blockName: 'Sample SingleBlock',
    });
    wrapper.update();
    wrapper.instance().handleEditBlock(event);
    expect(newProps.editBlock).toBeCalled();
  });

  it('should close the modal when the deleteBlock call fails and block is not provided', () => {
    wrapper.setProps({
      deleteBlock: jest.fn(() => Promise.reject()),
      block: {},
    });
    wrapper.update();
    wrapper.instance().handleDeleteBlock(event);
    expect(wrapper.state('closeModal')).toEqual(false);
  });

  it('should close the modal when the deleteBlock call fails and block is provided', () => {
    const error = new Error('There is an error while carrying out this operation');
    error.graphQLErrors = [{ message: ['error'] }];
    wrapper.setProps({
      deleteBlock: jest.fn(() => Promise.reject(error)),
      block: blockData,
    });
    wrapper.update();
    wrapper.instance().handleDeleteBlock(event);
    expect(instance.toggleLoading).toBeCalled();
    expect(wrapper.state('closeModal')).toEqual(false);
  });

  it('should resolve deleteBlock', () => {
    const newProps = {
      deleteBlock: jest.fn(() => Promise.resolve()),
      office: officeData,
      block: blockData,
      addBlock: jest.fn(),
      editBlock: jest.fn(),
    };
    wrapper = shallow(<BlockActions {...newProps} />);
    wrapper.instance().handleDeleteBlock(event);
    expect(newProps.deleteBlock).toBeCalled();
  });
});
