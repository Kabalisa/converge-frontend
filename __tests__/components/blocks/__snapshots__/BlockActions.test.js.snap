// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockActions Component renders properly 1`] = `
<BlockModal
  blockName="Container"
  closeModal={false}
  editing={false}
  handleCloseModal={[Function]}
  handleFormInputChange={[Function]}
  handleModalStateChange={[Function]}
  handleSubmit={[Function]}
  isLoading={false}
  office={
    Object {
      "id": "81",
      "location": Object {
        "name": "Kampala",
      },
      "name": "The Crest7",
    }
  }
/>
`;
